src/libraries/System.Private.CoreLib/src/System/SR.cs

```diff
@@ -13,48 +13,19 @@ namespace System
 {
     internal static partial class SR
     {
-        // This method is used to decide if we need to append the exception message parameters to the message when calling SR.Format.
-        // by default it returns false.
-        // Native code generators can replace the value this returns based on user input at the time of native code generation.
-        // Marked as NoInlining because if this is used in an AoT compiled app that is not compiled into a single file, the user
-        // could compile each module with a different setting for this. We want to make sure there's a consistent behavior
-        // that doesn't depend on which native module this method got inlined into.
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static bool UsingResourceKeys()
-        {
-            return false;
-        }
+        private static readonly object _lock = new object();
+        private static List<string>? _currentlyLoading;
+        private static int _infinitelyRecursingCount;
+        private static bool _resourceManagerInited;
 
         // Needed for debugger integration
         internal static string? GetResourceString(string resourceKey)
         {
             return GetResourceString(resourceKey, string.Empty);
         }
 
-        internal static string GetResourceString(string resourceKey, string? defaultString)
-        {
-            if (UsingResourceKeys())
-                return defaultString ?? resourceKey;
-
-            string? resourceString = null;
-            try { resourceString = InternalGetResourceString(resourceKey); }
-            catch (MissingManifestResourceException) { }
-
-            if (defaultString != null && resourceKey.Equals(resourceString, StringComparison.Ordinal))
-            {
-                return defaultString;
-            }
-
-            return resourceString!; // only null if missing resource
-        }
-
-        private static readonly object _lock = new object();
-        private static List<string>? _currentlyLoading;
-        private static int _infinitelyRecursingCount;
-        private static bool _resourceManagerInited = false;
-
         [PreserveDependency(".cctor()", "System.Resources.ResourceManager")]
```

Comment:
I'm a little confused by this attribute.  The method body uses a static method on ResourceManager: ResourceManager.GetString.  Why then without this attribute is the linker then able to remove ResourceManager's static cctor?

Reply:
This was here before the linker was able to recognize `RuntimeHelpers.RunClassConstructor`. I think it's better not to do linker clean as part of this PR

Reply:
> This was here before the linker was able to recognize RuntimeHelpers.RunClassConstructor

I'm not referring to RunClassConstructor.  I'm referring to the GetString call here:
https://github.com/dotnet/runtime/pull/36277/files/0b45fccee71a41af84cfdd32a5f32ea8cd7498a3#diff-263b7dc398fbc3b5e0c91e082bdd9bf4L125
Does calling a static method on a type not force the linker to keep that type's cctor?  If it doesn't, that seems really bad.

> I think it's better not to do linker clean as part of this PR

That's fine.  Can you please open an issue for it if there isn't one already?

Reply:
> Does calling a static method on a type not force the linker to keep that type's .cctor? If it doesn't, that seems bad.

The `ResourceManager` type is marked `beforefieldinit`, so the class constructor is only needed if there's an access to a static field - just calling a static method doesn't trigger it and it doesn't mean it's needed. Arguably, in this case if linker doesn't keep any of the static field accessing code, it would actually be fine to remove the class constructor too and have `RunClassConstructor` be a no-op. But linker now recognizes the `RunClassConstructor(typeof(Foo))` pattern, so this is an academic discussion at this point.

The annotation shouldn't be needed.

Reply:
Filles #36330